File: sdk/python/src/direnv_config/native.py
from __future__ import annotations

from pathlib import Path
from typing import Any

import yaml

from direnv_config.path import get_path


class NativeBackend:
    def __init__(self, store: Path) -> None:
        self._store = store

    def get(self, config: str, path: str | None = None) -> Any:
        active_file = self._store / config / ".active"
        if not active_file.exists():
            return None
        contents = active_file.read_text()
        data = yaml.safe_load(contents)
        if path is None:
            return data
        return get_path(data, path)

    def list_configs(self) -> list[str]:
        meta_file = self._store / ".meta"
        if not meta_file.exists():
            return []
        contents = meta_file.read_text()
        meta = yaml.safe_load(contents)
        if meta is None or not isinstance(meta, dict):
            return []
        configs = meta.get("configs", [])
        if not isinstance(configs, list):
            return []
        return [str(c) for c in configs]

File: sdk/elixir/lib/direnv_config/native.ex
defmodule DirenvConfig.Native do
  @spec get(String.t(), String.t(), String.t() | nil) :: {:ok, term()} | :error
  def get(store_path, config, path \\ nil) do
    active_file = Path.join([store_path, config, ".active"])

    case YamlElixir.read_from_file(active_file) do
      {:ok, data} ->
        resolve(data, path)

      {:error, _} ->
        :error
    end
  end

  @spec list_configs(String.t()) :: {:ok, [String.t()]} | :error
  def list_configs(store_path) do
    meta_file = Path.join(store_path, ".meta")

    case YamlElixir.read_from_file(meta_file) do
      {:ok, %{"configs" => configs}} when is_list(configs) ->
        {:ok, configs}

      _ ->
        :error
    end
  end

  defp resolve(data, nil), do: {:ok, data}
  defp resolve(data, path), do: DirenvConfig.Path.get(data, path)
end
